### PR TITLE
DOXYGEN/API: Addiging definitions for basic UCP concepts

### PR DESCRIPTION
**Pull request by shamisp**
I added IBM's copyright since I reused some of ideas
from PAMI documentation.

### CODE REVIEW

**Comment by shamisp** [comment]
autobot build

**Comment by shamisp** [comment]
@yosefe @sameerkm @sssharka please review

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1015/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/439/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Please review the documentation, I will address the compilation issue tomorrow.

**Comment by yosefe** [comment]
why did you remove the typedefs from ucp_def.h?

**Comment by shamisp** [comment]
it looks cleaner in terms of documentation

File: src/ucp/api/ucp_def.h [hunk 1]
```diff
@@ -13,17 +15,124 @@
...
+ * @ingroup UCP_CONTEXT
+ * @brief UCP Application Context
+ *
+ * UCP application context (or just a context) is an opaque handle that holds a
+ * UCP communication instance’s global information.  It represents a single UCP
+ * communication instance.  The communication instance could be an OS process
+ * (an application) that uses UCP library.  This global information includes
+ * communication resources, endpoints, memory, temporary file storage, and
+ * other communication information directly associated with a specific UCP
+ * instance.  The context also acts as an isolation mechanism, allowing
```
Comment (sssharka) [inline src/ucp/api/ucp_def.h]: We can add here that this allows for multiple independent communications capability

File: src/ucp/api/ucp_def.h [hunk 2]
```diff
@@ -13,17 +15,124 @@
...
+
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief UCP Endpoint
+ *
+ * The endpoint handle is an opaque object that is used to address a remote
+ * @ref ucp_worker_h "worker". It typically provides a description of source,
+ * destination, or both. All UCP communication routines address a destination
+ * with the endpoint handle. The endpoint handle is associated with only one
+ * @ref ucp_context_h "UPC context". UCP provides the @ref ucp_ep_create
```
Comment (sssharka) [inline src/ucp/api/ucp_def.h]: "UCP context" not UPC

File: src/ucp/api/ucp_def.h [hunk 3]
```diff
@@ -82,7 +191,5 @@ typedef void (*ucp_send_callback_t)(void *request, ucs_status_t status);
  *                        status is UCS_OK.
  */
 typedef void (*ucp_tag_recv_callback_t)(void *request, ucs_status_t status,
-                                        ucp_tag_recv_info_t *info);
-
-
+                                        struct ucp_tag_recv_info *info);
```
Comment (sssharka) [inline src/ucp/api/ucp_def.h]: why this change? ucp.h has the ucp_tag_recv_info_t..

File: src/ucp/api/ucp_def.h [hunk 4]
```diff
@@ -13,17 +15,124 @@
...
+ * tag matching semantics (derivative of MPI tag matching semantics).
+ */
 typedef uint64_t                         ucp_tag_t;
+
+/**
+ * @ingroup UCP_COMM
+ * @brief UCP Datatype Identifier
+ *
+ * UCP datatype identifier is a 64bit object used for datatype identification.
+ * Predefined UPC identifiers are defined by @ref ucp_type.
```
Comment (sssharka) [inline src/ucp/api/ucp_def.h]: UCP identifier not UPC

File: src/ucp/api/ucp_def.h (hunk 2 above)
Comment (shamisp) [inline src/ucp/api/ucp_def.h]: good catch

File: src/ucp/api/ucp_def.h (hunk 3 above)
Comment (shamisp) [inline src/ucp/api/ucp_def.h]: Will fix it

File: src/ucp/api/ucp_def.h (hunk 4 above)
Comment (shamisp) [inline src/ucp/api/ucp_def.h]: good catch

File: src/ucp/api/ucp_def.h (hunk 1 above)
Comment (shamisp) [inline src/ucp/api/ucp_def.h]: I think it is covered in the next sentence that provides an example of MPI and SHMEM

**Comment by shamisp** [comment]
I cleaned it a bit. Please review. I remove some unnecessary forward declarations.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1021/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/446/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1022/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/447/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/448/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Good to go ?

**Comment by yosefe** [comment]
:+1:
